Sign Up Feature

marias02 opened the pull request:
I added the sign up link, and the form and is working on the backend, but it's still not redirecting the user to the tweetes page, i'm still working on it! I added to the database the active_storage function, so a user can add a profile picture. I added some background to prevent errors from frontend input, in the birthdate and in the password...!

I've added the components, and the appropiate forms so the user is able to register. Even though I'm still working in the login functionality.

carlosarias1992 commented:
We don't need the ` .DS_Store` file here. We can remove it and add it to the `.gitignore` file, so that we don't merge it by accident.

carlosarias1992 commented on app/controllers/users_controller.rb:
```diff
@@ -31,22 +32,34 @@ def edit
-    user = @user
     @user.profile_picture.attach(user_params[:profile_picture])
     if @user.username === ""
       @user.username = @user.name + rand((User.all.length)..(User.all.length * 100000000)).to_s
```
What happens if the username we assign here has already been picked?

marias02 replied:
If the username has already been picked, then, the user stays with the username that he has already picked, this is just for preventing an error when creating the user, as the user cannot be saved without the username, but if the user doesn't pick any username, when registering, then a random unique username is made, based on the user name he has entered in the form, and a random number, just as in twitter. I didn't put any else statement, because i didn't see it necessary.

carlosarias1992 replied:
Let's say that user A joins retwitter and we assign them a username of `A1`, then user AA joins retwitter and we asssign them a user name of `A1` here. We end up with two users who have the same username because we're not checking if we already have another user with that username in the database.

marias02 replied:
If the user AA joins retwitter, then, the result username is going to be AA1. And, actually we are checking if the username is unique, as in the database i added the unique: true condition, so the username is not going to be saved if that username already exists.

carlosarias1992 replied:
So, if the username is not unique, when we try to save the username here, this will throw an error, right? We should check that the username doesn't exist already, and if it does, then we reassign a different username.

carlosarias1992 commented on app/controllers/users_controller.rb:
```diff
@@ -31,22 +32,34 @@ def edit
-    respond_to do |format|
-      if @user.save
+    if @user.birthdate.is_a? String
+        @user.birthdate = Date.parse(@user.birthdate)
```
We shouldn't pass the entire birthday as a string. This is fragile and prone to breaking. We should have different attributes for birthday month, day, and year. Month should preferably be a number as well.

marias02 replied:
So, you say we should pass the month, day and year, as separated columns in the database as well, or just in the frontend?

carlosarias1992 replied:
We can store the date in one column in the backend, but the frontend should give us the date separate.

carlosarias1992 commented on app/javascript/components/SignUp.jsx:
```diff
@@ -30,28 +32,39 @@ class SignUp extends Component {
 
     async formSubmit(formData){
-        var data = new FormData(formData);
+        console.log(this.state)
```
Please remove this `console.log`

carlosarias1992 commented on app/javascript/components/SignUp.jsx:
```diff
@@ -1,4 +1,5 @@
 import React, { Component } from "react";
+import Tweetes from '../components/Tweetes';
```
This component should be renamed to `Tweets` instead of `Tweetes`

marias02 replied:
I put this name to make it a little bit different from the original tweets, from Twitter, as this website is called, retwitter, i changed the names of the things it contains, to make it different as well.

carlosarias1992 replied:
That's not necessary. This name is confusing and it will get harder to follow in the future if we keep changing the names.

carlosarias1992 commented on app/javascript/packs/application.js:
```diff
@@ -15,3 +15,25 @@ require("channels")
 // const images = require.context('../images', true)
 // const imagePath = (name) => images(name, true)
+
+const app = require('express')();
```
Why are we using express here? Express is relatively the same thing as Ruby on Rails. We don't need it to make the app work.

carlosarias1992 commented on app/javascript/components/Tweetes.jsx:
```diff
@@ -19,9 +19,6 @@ class Tweetes extends Component {
             })
             .then(response => this.setState({ tweetes: response }))
             .catch(() => this.props.history.push("/"));
-            // fetch(url)
```
Very good! Commented out code leads to noise. As more people contribute to the same code base, commented out code tends to persist because nobody knows if the commented out code is serving a purpose. Always try to remove commented out code if you don't need it anymore.

carlosarias1992 commented on app/javascript/components/SignUp.jsx:
```diff
@@ -30,28 +32,39 @@ class SignUp extends Component {
         return(
             <form onSubmit={ this.handleSubmit }>
-                <input type="text" name="name" value={ this.state.name } placeholder="Name"
+                <label htmlFor="name">Name: </label>
```
nit: I see a lot of duplication here. We could reduce the duplicates by creating a reusable component for inputs. For example: 

```
function Input({ label, htmlFor, inputID, name, value, onChange, placeholder, type, required }) {
   return (
      <label htmlFor={ htmlFor }>{ label }</label>
      <input type={ type } id={ inputID } name={ name } value={ value } onChange={onChange} placeholder={ placeholder } required={ required } />
   )
}
```

marias02 commented:
I have fixed the the errors found in the backend, as it showed in the comments, i have fixed the frontend, so now the user enters is birthdate in separate columns, and sends it to the backend, in one value. 

I have also changed the name of Tweetes to Tweets, and now the user can add a profile picture, is properly registered in the backend, and is redirected to the tweets page, after a successful response from the backend.

carlosarias1992 commented on app/javascript/components/Login.jsx:
```diff
@@ -0,0 +1,48 @@
+        return(
+            <div>
+                <form onSubmit={ this.onSubmitForm } id="ajax_form" className="login">
+                    <label htmlFor="username">Username: </label>
```
Use `Input` here

carlosarias1992 commented on app/javascript/components/SignUp.jsx:
```diff
@@ -1,4 +1,23 @@
+import { Redirect } from "react-router";
+
+// i'm gonna change this approach when i see it working
+const days = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17,
```
Consider using a library for these values

carlosarias1992 commented on app/controllers/users_controller.rb:
```diff
@@ -31,22 +32,36 @@ def edit
     @user.profile_picture.attach(user_params[:profile_picture])
     if @user.username === ""
-      @user.username = @user.name + rand((User.all.length)..(User.all.length * 100000000)).to_s
+      while !User.find_by(username: @user.username){ 
```
Let's manually test this to make sure it's working as expected

carlosarias1992 commented on app/controllers/users_controller.rb:
```diff
@@ -31,22 +32,36 @@ def edit
-    respond_to do |format|
-      if @user.save
+    if @user.birthdate.is_a? String
+        @user.birthdate = Date.parse(@user.birthdate)
```
Let's investigate to see if we can pass custom attributes from the frontend

carlosarias1992 commented on app/javascript/components/SignUp.jsx:
```diff
@@ -29,29 +53,78 @@ class SignUp extends Component {
+        });
+    }
+
+    onChangeBdayOpt(e){
```
It's better to be more verbose. We could lose the meaning eventually and it will be hard to know what this function was intended for. Let's change it to `onChangeBirthdateOption`.

marias02 commented:
As requested, i changed the name of the function from onChangeBdayOpt to onChangeBirthdateOption, so now it is understandable, also i tested the function to make sure it was working as expected,  i removed some bugs that were making the function unable to work properly. Also, i couldn't find any library that makes the markdowns dates easier to implement.

And when you finish checking these changes, i have some questions for you regarding the requested change, for passing custom values from the frontend to the backend.

carlosarias1992 commented on app/controllers/users_controller.rb:
```diff
@@ -31,20 +32,45 @@ def edit
-        render json: user
+    if @user.birthdate.is_a? String
+      @user.birthdate = Date.parse(@user.birthdate)
+      if Date.valid_date?(@user.birthdate.month, @user.birthdate.day, @user.birthdate.year)
```
Nice! I like this

carlosarias1992 commented on app/controllers/users_controller.rb:
```diff
@@ -31,20 +32,45 @@ def edit
+    if @user.username == ""
+      while User.find_by(username: @user.username) && @user.username == ""
+        @user.username = @user.name + rand((User.all.length)..(User.all.length * 100000000)).to_s
+        @user.save
```
Why are we using so many `@user.save`? I see one in lines 39, 45, 51, 60, and 63. I believe we just need the one in line 63 to make sure we can save the users to the db without any errors.

carlosarias1992 commented on app/javascript/functions/BirthMonth.jsx:
```diff
@@ -0,0 +1,21 @@
+
+    months.push(<option name={optionName} key="0.1" value="" selected disabled></option>)
+
+    for (let i = 0; i < monthsArr.length; i++) {
```
We could simplify this for loop by using a forEach here. For example: 

```
monthsArr.forEach((month, i) => months.push(option key={i.toString()} name={optionName} value={month}>{month}</option>))
```

carlosarias1992 commented on yarn.lock:
```diff
@@ -3,21 +3,21 @@
 
```
Why do we have a `yarn.lock` and a `package-lock.json`? We should just use one of the two, but not both. https://stackoverflow.com/a/46590427

marias02 replied:
If i remove the file solves the problem? Or do i need to do something else?

carlosarias1992 replied:
Yes, you can remove one of the two

carlosarias1992 commented on app/javascript/components/SignUp.jsx:
```diff
@@ -23,35 +32,70 @@ class SignUp extends Component {
+    async formSubmit(){
         await fetch('/signup_user', {
-            Accept: 'application/json',
+            headers: {
```
We should create a utils folder for api call functions. The more api calls we have to make, the messier this will get. ie. https://github.com/carlosarias1992/Fakebook/blob/61de622171724ea1efc366049ace95150ebfd081/frontend/util/session_api_util.js#L1

carlosarias1992 replied:
By the way, you don't have to use ajax, but that's an example of the util function for signing up users.

marias02 commented:
I have done the changes that you have requested! Happy reviewing!

carlosarias1992 reviewed (approved):
Nice work!
